sageresearch-app-sdk/src/main/java/org/sagebionetworks/research/sageresearch/viewmodel/ScheduleRepository.kt

```diff
@@ -176,17 +229,207 @@ open class ScheduleRepository(context: Context) {
+     *                 schedule can be null if the user does a task without internet
+     * @param taskResult the result of the user doing the schedule task
+     */
+    fun updateSchedule(taskResult: TaskResult) {
+        // If we previously registered a guid with a taskRunUuid, we should be able to find the schedule in the
+        // the database at this point based on the schedule guid
+        findSchedule(taskResult.taskUUID, Action1 { schedule ->
+            // TODO: mdephillips 9/14/2018 in past apps, here we would set client data from TaskResult
+            // TODO: mdephillips 9/14/2018 instead use TaskResult to generate study report for schedule
+            schedule.startedOn = taskResult.startTime
+            schedule.finishedOn = taskResult.endTime
+            cacheSchedule(schedule)
+            updateScheduleToBridge(schedule)
+        }, Action1 {
+            // TODO: mdephillips 9/14/2018 message the user there will be no history?
+            logger.warn(it.localizedMessage)
+        })
+    }
+
+    /**
+     * Private function should only be accessed through this class as a re-usable way to update
+     * a schedule on bridge.
+     */
+    private fun updateScheduleToBridge(schedule: ScheduledActivityEntity) {
+        val bridgeSchedule = schedule.clientWritableCopy()
+        updateActivityOnBridge(bridgeSchedule, Action1 {
+            schedule.needsSyncedToBridge = false
+            cacheSchedule(schedule)
+        }, Action1 {
+            scheduleUpdateFailed(schedule, it)
+        })
+    }
+
+    /**
+     * Encapsulate db call functionality into it own function for providing custom mock behavior in tests,
+     * and also to remove redundant threading code throughout the class.
+     */
+    @VisibleForTesting
+    open fun findSchedule(
+            taskRunUuid: UUID,
+            onNext: Action1<ScheduledActivityEntity>,
+            onError: Action1<Throwable>) {
+
+        // Null schedule usually means that the developer forgot to call createScheduleTaskRunUuid before launching
+        // the task associated with the schedule.
+        // Or the user is offline and did a task without having synced the study's schedules.
+        // Which, that is okay, the s3 upload will still take place, but we won't be able to update the schedule on bridge.
+        // Side effects for this will be, no study reports or finishedOn status for the task will show in the UI.
+        // TODO: mdephillips 9/14/2018 message the user there will be no history?
+        val guid = scheduleTaskRunUuidMap[taskRunUuid] ?: run {
+            onError.call(Throwable("No schedule guid found for taskRunUuid $taskRunUuid, " +
+                    "are you sure you function createScheduleTaskRunUuid() before running the task?"))
+            return
+        }
+
+        scheduleRepoSubscriptions.add(Observable.just(scheduleDao)
+                .subscribeOn(Schedulers.io())
+                .subscribe({ dao ->
```

Comment:
I think you asked about this in the old PR
```.subscribeOn(Schedulers.io())
.observeOn(AndroidSchedulers.mainThread())
.subscribe( // actions called on main thread
onNextAction,
onErrorAction
);

Reply:
`subscribeOn` is a bit confusing because it doesn't control the thread where `subscribe` actions get called. Rather, it controllers the thread the `Subscription` observes on _when_ `subscribe` is called. So `subscribeOn(Schedulers.io())` means the original Observable starts off on the io thread.